homeassistant/components/sonos/media_browser.py

```diff
@@ -53,15 +53,18 @@ def get_thumbnail_url_full(
     media_content_type: str,
     media_content_id: str,
     media_image_id: str | None = None,
+    item: MusicServiceItem | None = None,
 ) -> str | None:
     """Get thumbnail URL."""
     if is_internal:
-        item = get_media(
-            media.library,
-            media_content_id,
-            media_content_type,
-        )
-        return urllib.parse.unquote(getattr(item, "album_art_uri", ""))
+        if not item:
+            item = get_media(
+                media.library,
+                media_content_id,
+                media_content_type,
+            )
+        album_art: str = urllib.parse.unquote(getattr(item, "album_art_uri", ""))
+        return album_art.replace("'", "%27").replace("(", "%28").replace(")", "%29")
```

Comment:
Why do we need to encode these values manually?
It does not feel right and can be error-prone. Do we have an alternative?

Reply:
Great question.  

Sonos is returning this string from the getattr(item, "album_art_uri", "").  This URL is double encoded.  However the "'" in a Hard Day's Night is not encoded at all.  The same is true for uri that have the open or close parenthesis in them. The presence of these characters causes the media browser to not display the artwork.

```
http://192.168.42.2:1400/getaa?u=x-file-cifs%3a%2f%2f192.168.42.100%2fmusic%2fThe%2520Beatles%2fA%2520Hard%2520Day's%2520Night%2f01%2520A%2520Hard%2520Day's%2520Night%25201.m4a&v=53
```

As far as I can tell; these are the only characters needed - at least in my 300+ albums.  I've compared the Sonos Browser side by side with the HA media browser and it is consistent.

Adding a code comment could be a good idea?

Reply:
I just tried that and a few variations:

The URI is composed of two parts:

The call to the speaker REST api to get the artwork
```
http://192.168.42.2:1400/getaa?
```

And the two URL parameters (u and v)

```
u=x-file-cifs%3a%2f%2f192.168.42.100%2fmusic%2fThe%2520Beatles%2fA%2520Hard%2520Day's%2520Night%2f01%2520A%2520Hard%2520Day's%2520Night%25201.m4a&v=53
```

After the first decode it looks like this

```
u=x-file-cifs://192.168.42.100/music/The%20Beatles/A%20Hard%20Day%27s%20Night/01%20A%20Hard%20Day's%20Night%201.m4a&v=53
```

There are two parameters u and v, and we'd need to split these parameters out - specifically the "u" parameter - URL decode it; then url encode it; and then re-assemble the whole call.  Because if we just URL decode and encode it; the "u=" and "v=" get URL encoded.  Anyways, this is starting to look fairly complicated and get's the code tied to the format of the URL.

Reply:
I added a detailed comment into the code. I also re-reviewed the soco API, and verified it is not constructing this URI and is just returning what the speaker returns.

Reply:
in order to simplify reviewing, I have removed the code to fix the bad image urls and will put that in a separate pull request. This pull request will now only address the performance issue and add the supporting missing tests.